### PR TITLE
Adding functionality to pin ansible-lint versions through input variable

### PR DESCRIPTION
**Pull request by xUnholy**
Fixes: #16 
- ~Addressed some linting concerns picked up by [shellcheck](https://github.com/koalaman/shellcheck)~
- ~Changed `ADD` to `COPY` as recommended by docker as ADD has additional functionality that may not be intended.~
- Allow new version input so the user can pin their version to align with local development

### CODE REVIEW

**Comment by xUnholy** [comment]
@stoe someone with write access will have to perform the merge as I don't seem to have access.

**Comment by stoe** [comment]
> someone with write access will have to perform the merge as I don't seem to have access

@webknjaz, do you want to do the honors?

File: entrypoint.sh [hunk 1]
```diff
@@ -79,17 +86,19 @@ parse_args() {
 ansible::lint() {
   : "${TARGETS?No targets to check. Nothing to do.}"
   : "${GITHUB_WORKSPACE?GITHUB_WORKSPACE has to be set. Did you use the actions/checkout action?}"
-  pushd ${GITHUB_WORKSPACE}
+  : "${VERSION?No version set. Default: ${DEFAULT_VERSION}}"
```
Comment (DO1JLR) [inline entrypoint.sh]: Hi,

As far as I understand if you add new variables you have to specify them in the action.yml file to make github actions use them.

Without it it is not working [in my github-action check](https://github.com/roles-ansible/role_ranger/runs/538944228). *-> [example git commit](https://github.com/roles-ansible/role_ranger/blob/cda67e7ceb75d6696c2faf4c991102ed23c3a554/.github/workflows/ansible-linting-chek.yml)*

File: entrypoint.sh [hunk 2]
```diff
@@ -79,17 +86,19 @@ parse_args() {
...
+  pushd "${GITHUB_WORKSPACE}"
 
-  local opts=$(parse_args "$@" || exit 1)
+  install_ansible_lint "${VERSION:-${DEFAULT_VERSION}}"
+  local opts
+  opts=$(parse_args "$@" || exit 1)
 
-  ansible-lint -v --force-color $opts ${TARGETS}
+  ansible-lint -v --force-color "$opts" "${TARGETS}"
```
Comment (DO1JLR) [inline entrypoint.sh]: Hi,

Usually I like adding quotation marks.
But is it realy a good idea adding them at the ``$opts`` variable?

In my testing case this variable is empty and will cause trouble.

File: entrypoint.sh (hunk 1 above)
Comment (stoe) [reply entrypoint.sh]: Good point @DO1JLR, I missed that one in my review.

File: entrypoint.sh (hunk 2 above)
Comment (webknjaz) [reply entrypoint.sh]: Valid point

Comment (xUnholy) [reply entrypoint.sh]: Should this be addressed that there isn't sufficient logic rather than remove the quoting and potentially have globbing? Easy to undo quotes just seems like there should be a more robust way to ensure it's used safely.

File: entrypoint.sh (hunk 1 above)
Comment (xUnholy) [reply entrypoint.sh]: Updated, thanks for that catch 👍

File: entrypoint.sh (hunk 2 above)
Comment (xUnholy) [reply entrypoint.sh]: Added extra validation to ensure if there was an empty `opts` string that it wouldn't be placed into the cmd, also tested the cmd with and without the validation and can confirm that the ansible-lint tool doesn't seem to be able to handle the empty string, even whether it was before or after the TARGET path.

File: action.yml [hunk 3]
```diff
@@ -34,10 +34,15 @@ inputs:
...
+      version in order to maintain consistency with local development.
+    required: false
 runs:
   using: docker
   image: Dockerfile
   args:
-  - ${{ inputs.args }}
+    - ${{ inputs.args }}
   env:
     TARGETS: ${{ inputs.targets }}
```
Comment (DO1JLR) [inline action.yml]: Hi,

My [ansible test](https://github.com/roles-ansible/role_ranger/runs/539358245) *([-> source](https://github.com/roles-ansible/role_ranger/blob/b4ad0c4967d26b690daf322f98176652cd7a953b/.github/workflows/ansible-linting-chek.yml))* is still failing with this message:
```bash
/entrypoint.sh: line 89: VERSION: No version set. Default: 4.2.0
```

You have to add the VERSION variable like in [this](https://github.com/xUnholy/ansible-lint-action/pull/2) PR:
```
env:
  TARGETS: ${{ inputs.targets }}
  VERSION: ${{ inputs.version }}
```

File: Dockerfile [hunk 4]
```diff
@@ -9,7 +9,5 @@ LABEL "com.github.actions.description"="Run Ansible Lint"
 LABEL "com.github.actions.icon"="activity"
 LABEL "com.github.actions.color"="gray-dark"
 
-RUN pip install ansible-lint
```
Comment (webknjaz) [inline Dockerfile]: I'm pretty sure that this should remain with some pinned version for users relying on tags. We should just install deps on top later.

File: Dockerfile [hunk 5]
```diff
@@ -9,7 +9,5 @@ LABEL "com.github.actions.description"="Run Ansible Lint"
...
-
-ADD entrypoint.sh /entrypoint.sh
```
Comment (webknjaz) [inline Dockerfile]: Why are you changing this? I usually reject unrelated changes in PRs as they block the review even if everything else is ok.

File: Dockerfile (hunk 4 above)
Comment (webknjaz) [reply Dockerfile]: I pin it here when releasing concrete tags: https://github.com/ansible/ansible-lint-action/commit/f8608f547bbe585e82634c54859e170c39384aa4.

File: action.yml [hunk 6]
```diff
@@ -2,7 +2,7 @@ name: Ansible Lint
 description: Run Ansible Lint
 author: Ansible by Red Hat <info@ansible.com>
 
-inputs:
```
Comment (webknjaz) [inline action.yml]: this change is unrelated, submit separately.

File: action.yml [hunk 7]
```diff
@@ -34,10 +34,15 @@ inputs:
...
+    description: |
+      Version of ansible-lint that should be used. This allows to pin to a target
+      version in order to maintain consistency with local development.
+    required: false
 runs:
   using: docker
   image: Dockerfile
   args:
-  - ${{ inputs.args }}
+    - ${{ inputs.args }}
```
Comment (webknjaz) [inline action.yml]: don't reindent things. it's unrelated and wrong. it was just as I wanted it before.

File: entrypoint.sh [hunk 8]
```diff
@@ -54,7 +57,7 @@ parse_args() {
         shift
         break
         ;;
-      -*|--*=) # unsupported flags
```
Comment (webknjaz) [inline entrypoint.sh]: This is unrelated to adding a version. It shouldn't be submitted in this PR.

File: action.yml [hunk 9]
```diff
@@ -34,10 +34,15 @@ inputs:
       or valid Ansible directories according to the Ansible role
       directory structure.
     required: true
+  version:
```
Comment (webknjaz) [inline action.yml]: rather than accepting just ansible-lint version, we should accept a list of python dists, in case if the users use some extra plugins and they are needed in the container. this would also address #8.

File: entrypoint.sh [hunk 10]
```diff
@@ -71,6 +74,10 @@ parse_args() {
   return 0
 }
 
+install_ansible_lint() {
+  pip install ansible-lint=="$1" || exit 1
```
Comment (webknjaz) [inline entrypoint.sh]: I'm pretty sure that the second part is unnecessary because of the `set` flags at the top of this script. Also, as per my version input comment, we shouldn't try to only accept the version, only the dists list.

File: entrypoint.sh [hunk 11]
```diff
@@ -79,17 +86,24 @@ parse_args() {
...
-  ansible-lint -v --force-color $opts ${TARGETS}
+  install_ansible_lint "${VERSION:-${DEFAULT_VERSION}}"
+  local opts
+  opts=$(parse_args "$@" || exit 1)
+  if [[ -z "$opts" ]]; then
+    # If no opts are provided an empty string causes ansible-lint to crash.
+    ansible-lint -v --force-color "${TARGETS}"
+  else
+    ansible-lint -v --force-color "$opts" "${TARGETS}"
+  fi
```
Comment (webknjaz) [inline entrypoint.sh]: this change is wrong. it should be possible to pass multiple args/opts/targets. it should be reverted.

File: entrypoint.sh [hunk 12]
```diff
@@ -79,17 +86,24 @@ parse_args() {
...
+  
 }
 
-
 args=("$@")
 
-if [ "$0" = "$BASH_SOURCE" ] ; then
+if [ "$0" = "${BASH_SOURCE[*]}" ] ; then
   >&2 echo -E "\nRunning Ansible Lint...\n"
-  ansible::lint ${args[@]}
```
Comment (webknjaz) [inline entrypoint.sh]: If you want to do some style/linter changes, put them into a separate PR.

File: Dockerfile (hunk 4 above)
Comment (xUnholy) [reply Dockerfile]: Why would you bother reinstalling the same dep - you can still use tags by simply updating the default version, this would then just be redundant.

File: Dockerfile (hunk 5 above)
Comment (xUnholy) [reply Dockerfile]: `ADD` is the wrong operation in this context.

"COPY and ADD are both Dockerfile instructions that serve similar purposes. They let you copy files from a specific location into a Docker image. COPY takes in a src and destination. ... A valid use case for ADD is when you want to extract a local tar file into a specific directory in your Docker image"

From what I can see this is not doing anything in relation to extracting files into the dir, in this case, `COPY` would be best suited.

File: action.yml (hunk 6 above)
Comment (xUnholy) [reply action.yml]: Whitespace my linter found surely doesn't warrant a new pull request.

File: action.yml (hunk 9 above)
Comment (xUnholy) [reply action.yml]: Happy to take a look at this issue independently to this one - I agree that would be a good idea.

File: action.yml (hunk 7 above)
Comment (xUnholy) [reply action.yml]: Pragmatically there should be some form of lint style file so others know what "just as I wanted" means.

This is a YAML linting change, unrelated sure, however, there was no default so I don't think we can argue this point. I will revert if this is the consensus.

File: action.yml (hunk 3 above)
Comment (xUnholy) [reply action.yml]: Thanks @DO1JLR  - I seemed to have missed that one, good catch!

File: entrypoint.sh (hunk 8 above)
Comment (xUnholy) [reply entrypoint.sh]: It's a redundant piece of code and logic, see the commit message for why. I would suggest a CI task that validates and checks these scripts to ensure dirty code isn't introduced in the future.

File: entrypoint.sh (hunk 10 above)
Comment (xUnholy) [reply entrypoint.sh]: It's just a force of habit cause people like to change the `set` flags and then the script should be robust enough to still operate the same.

File: entrypoint.sh (hunk 11 above)
Comment (xUnholy) [reply entrypoint.sh]: I'm not sure what this feedback actually means? Read my commit messages about why $opts was quoted, then read @DO1JLR  feedback on how that is expanded and breaks the ansible-lint CLI if it's an empty string. This encapsulated the variable with quotes as strongly recommended by almost every linter and shellchecker you'll find, and also allow for the string to not be put into the command if it's empty.

File: action.yml (hunk 6 above)
Comment (xUnholy) [reply action.yml]: @DO1JLR  has been kind enough to address this on his PR - https://github.com/ansible/ansible-lint-action/pull/18

File: Dockerfile (hunk 4 above)
Comment (webknjaz) [reply Dockerfile]: Because there's no version sourcing from Git tags here and I want to have a default bundled at build time.

File: Dockerfile (hunk 5 above)
Comment (webknjaz) [reply Dockerfile]: Both work, `ADD` just has more features. But either way, it's not a change that belongs to this PR.

File: action.yml (hunk 7 above)
Comment (webknjaz) [reply action.yml]: I've added a yamllint config to master.

File: entrypoint.sh (hunk 8 above)
Comment (webknjaz) [reply entrypoint.sh]: Sure but it's not related to this specific PR. Such a change should be submitted along with tests/CI configuration separately.

File: entrypoint.sh (hunk 11 above)
Comment (webknjaz) [reply entrypoint.sh]: Will it work if `opts="--arg1 --arg2"`? Will these args be interpolated separately? I'm pretty sure that everything quoted will be passed as a single arg.

File: action.yml (hunk 9 above)
Comment (webknjaz) [reply action.yml]: I'm pretty sure that this PR is only a special case of #8 and I wouldn't want to remake the API every release.

File: action.yml (hunk 7 above)
Comment (xUnholy) [reply action.yml]: Will pull master and rerun the linter, thank you for adding that 👍

File: entrypoint.sh (hunk 11 above)
Comment (xUnholy) [reply entrypoint.sh]: Reverted. I would suggest adding a CI linter and testing issuer that can capture some of these edge cases where I could break the script by passing broken args, will leave that to your discretion.

File: Dockerfile (hunk 4 above)
Comment (xUnholy) [reply Dockerfile]: The variable here can be used as your default https://github.com/xUnholy/ansible-lint-action/blob/05c999845d321220371afa4a9acc463240b1f0d4/entrypoint.sh#L7

File: action.yml (hunk 9 above)
Comment (xUnholy) [reply action.yml]: Looked at some other projects, I would keep the version uniquely set for simplicity of purpose to the users. #8 is an edge case that could be its own input variable for more advanced users that have this requirement.

**Comment by xUnholy** [comment]
@webknjaz have addressed your feedback either with changes and/or comments - if you could please re-review that would be fantastic 👍

**Comment by DO1JLR** [comment]
Hi @xUnholy 

I tested your PR again and my github action task is still failing. Sorry :disappointed: 
I hope you see the error and will resolve it :crossed_fingers: 

My Task: [.github/workflows/ansible-linting-chek.yml](https://github.com/roles-ansible/role_ranger/blob/3d70bacf7793e8a4fdf5aab207eab23aae109356/.github/workflows/ansible-linting-chek.yml)
Error Log: [github action log (runs/551832290)](https://github.com/roles-ansible/role_ranger/runs/551832290)
Error Message:
```
Lint Ansible Playbook
3s
+ exit 1
/usr/bin/docker run --name e87b52a02ea588f7e54278a1646ad708fbdc64_b345c3 --label e87b52 --workdir /github/workspace --rm -e INPUT_TARGETS -e INPUT_ARGS -e INPUT_VERSION -e TARGETS -e VERSION -e HOME -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e RUNNER_OS -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/role_ranger/role_ranger":"/github/workspace" e87b52:a02ea588f7e54278a1646ad708fbdc64  ""
/github/workspace /github/workspace
+ DEFAULT_VERSION=4.2.0
+ args=("$@")
+ '[' /entrypoint.sh = /entrypoint.sh ']'
+ echo -E '\nRunning Ansible Lint...\n'
\nRunning Ansible Lint...\n
+ ansible::lint ''
+ : .
+ : /github/workspace
+ pushd /github/workspace
+ install_ansible_lint .
+ pip install ansible-lint==.
WARNING: The directory '/github/home/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
ERROR: Could not find a version that satisfies the requirement ansible-lint==. (from versions: 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 2.0.1, 2.0.3, 2.1.0, 2.1.3, 2.2.0, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.5, 2.3.6, 2.3.8, 2.3.9, 2.4.0, 2.4.1, 2.4.2, 2.5.0, 2.6.1, 2.6.2, 2.7.0rc1, 2.7.0, 2.7.1, 3.0.0rc1, 3.0.0rc2, 3.0.0rc3, 3.0.0rc4, 3.0.0rc5, 3.0.0rc6, 3.0.0rc7, 3.0.0rc8, 3.0.0rc9, 3.0.0rc10, 3.0.0rc11, 3.0.0rc12, 3.0.0, 3.0.1, 3.0.2rc1, 3.1.0rc1, 3.1.0, 3.1.1, 3.1.2, 3.1.3, 3.2.0, 3.2.1rc1, 3.2.1rc2, 3.2.1, 3.2.2rc1, 3.2.2, 3.2.3, 3.2.4, 3.2.5, 3.3.0rc1, 3.3.0rc2, 3.3.0rc3, 3.3.0rc4, 3.3.0, 3.3.2, 3.3.3, 3.4.0rc1, 3.4.0rc2, 3.4.0, 3.4.1, 3.4.3, 3.4.4, 3.4.5, 3.4.6, 3.4.7, 3.4.8, 3.4.9, 3.4.10, 3.4.11, 3.4.12, 3.4.13, 3.4.15, 3.4.16, 3.4.17, 3.4.18, 3.4.19, 3.4.20, 3.4.21, 3.4.22, 3.4.23, 3.5.0rc1, 3.5.0, 3.5.1, 4.0.0a1, 4.0.0a2, 4.0.0, 4.0.1, 4.1.0a0, 4.1.0, 4.1.1a0, 4.1.1a1, 4.1.1a2, 4.1.1a3, 4.1.1a4, 4.1.1a5, 4.1.1a6, 4.2.0a1, 4.2.0rc1, 4.2.0rc2, 4.2.0)
ERROR: No matching distribution found for ansible-lint==.
+ exit 1
```

File: action.yml (hunk 9 above)
Comment (webknjaz) [reply action.yml:37]: I'm still not sure if I want to accept the PR with a separate version for only ansible-lint.

**Comment by xUnholy** [comment]
@DO1JLR Thanks for the quick test file - found the issue, I didn't save my changes on the inputs so it was passing `targets` to `version` input, correct this now. 👍 

Once this gets merged I'll add a test workflow so we have some validation during CI if proposed changes will affect any of existing functionality @webknjaz.

**Comment by webknjaz** [comment]
I still feel strongly about https://github.com/ansible/ansible-lint-action/pull/17#discussion_r399290361

**Comment by xUnholy** [comment]
> I still feel strongly about [#17 (comment)](https://github.com/ansible/ansible-lint-action/pull/17#discussion_r399290361)

I think that is definitely a valid point, however, we can allow users to explicitly set versions of the CLI with the "version" parameter and create a more generic field, that can pass multiple args into a `pip install` logic.

I'm also probably not across enough of the detail in terms of #8 because I'd have no way to test it, and I'm not familiar with what he's trying to accomplish, to be honest.

**Comment by webknjaz** [comment]
> I'm also probably not across enough of the detail in terms of #8 because I'd have no way to test it, and I'm not familiar with what he's trying to accomplish, to be honest.

That issue basically only needs an extra run of `pip install` with the provided dist specs, that's it. This is also #19.

> I think that is definitely a valid point, however, we can allow users to explicitly set versions of the CLI with the "version" parameter and create a more generic field, that can pass multiple args into a `pip install` logic.

In GitHub Actions land, it's common to use a repo tag for this (that it's *also* how things are versioned in Marketplace). Adding an extra version param would be confusing:
```
- uses: ansible/ansible-lint-action@4.1.0
  with:
    version: 3.1.0
```
How is this readable?

Besides, we *need* to have the version matching the tag pre-installed during the build step. If we were to publish the container to a docker registry, it'd have a ready-to-use container w/o extra time spent in runtime.

I also don't want to overcomplicate the script logic.

**Comment by xUnholy** [comment]
All valid points, I'll try my best to address them for you.

> That issue basically only needs an extra run of pip install with the provided dist specs, that's it. This is also #19.

Correct, I'm sure in theory it's a trivial change, it just might be suited better for someone who has a valid use case to test it within the context of using other pip dependencies for ansible-lint.
I think this behaviour is not how I'd want to explain to a user that this is the method for explicitly setting a version.

> In GitHub Actions land, it's common to use a repo tag for this (that it's also how things are versioned in Marketplace). Adding an extra version param would be confusing:

This is true actually of both scenarios, I'm used to more trunk-based development so my view is that master should always be referenced and no breaking changes ever get introduced into the GitHub action and it would be safe for users to always point to master, it becomes annoying if users have to update to newer tagged releases just to get new functionality. If we want to make something that would be deemed a breaking change, or logic change, it should become its own new input type, exactly the same philosophy for API development.

This would remove the issue of it being "confusing" or "un-readable". Example:

```
- uses: ansible/ansible-lint-action@master
  with:
    version: 3.1.0
```

If you were to introduce your extra dependency input I would consider it being something like this:

```
- uses: ansible/ansible-lint-action@master
  with:
    targets: .
    version: 3.1.0
    dep(endencies): x==1,y==2,z==3
```

If we didn't have a version as an explicit input the user would interface with your action like this:

```
- uses: ansible/ansible-lint-action@master
  with:
    targets: .
    dep(endencies): ansible-lint==4.2.0
```

> I also don't want to overcomplicate the script logic.

This is always complicated to rebut because "complicated" is very subjective to the user - I would hope with validation, and testing that a user never needs to look at the script logic, and we never introduce anything that would be deemed a breaking change.

This change really needs to become a little more collaborative, everyone has strong opinions on the best ways to do things, I'd love to hear others thoughts on the matter if @DO1JLR @stoe

**Comment by webknjaz** [comment]
> I'm used to more trunk-based development so my view is that master should always be referenced and no breaking changes ever get introduced into the GitHub action and it would be safe for users to always point to master, it becomes annoying if users have to update to newer tagged releases just to get new functionality

This is not how the current semantics works. It's more deterministic when there's some pinned pointer in place.
And don't even get me started on moving ~targets~ Git pointers. This is a total security hazard from the user perspective: https://julienrenaux.fr/2019/12/20/github-actions-security-risk/. So I'd not even recommend users use tags from this PoV.

> This should also be argued that it is "confusing" and "un-readable" right?

I was thinking along the lines of
```yaml
- uses: ansible/ansible-lint-action@4.1.0
  with:
    targets: .
    override-deps: |
      ansible==2.10
      ansible-lint==4.2.0
```
which communicates the intent of the input in a more clear fashion.

**Comment by xUnholy** [comment]
> This is not how the current semantics works. It's more deterministic when there's some pinned pointer in place.
And don't even get me started on moving targets Git pointers. This is a total security hazard from the user perspective: https://julienrenaux.fr/2019/12/20/github-actions-security-risk/. So I'd not even recommend users use tags from this PoV.

His opening line says using tags may lead to the same problem? I'm not actually sure what point you're trying to make here?

> Malicious code can be inserted into any GitHub action, even those which are tagged.

```
- uses: ansible/ansible-lint-action@4.1.0
  with:
    targets: .
    override-deps: |
      ansible==2.10
      ansible-lint==4.2.0
```

Compared to the proposal:

```
- uses: ansible/ansible-lint-action@master
  with:
    targets: .
    version: 4.2.0
```

The action can be agnostic to version this way and the user is just explicitly determining a pinned version or they're happy with whatever the default is provided by the Github action. In fact, they could still use a tag if that is what you really want to commit to doing in the future and maintaining, in which they'd just do as follows:

```
- uses: ansible/ansible-lint-action@4.2.0
  with:
    targets: .
```

In which the entrypoint.sh `DEFAULT_VERSION` gets bumped with the tag you release.

**Comment by xUnholy** [comment]
If you like I'm happy for you to close this pull request, I appreciate everyone's hard work on this project, I might just maintain my own fork for my use case to simplify version pinning, I wasn't really prepared to spend too long debating on this topic and then having to implement additional feature requirements that were not to solve my original issue.

Point of reference if anyone was following this thread: https://github.com/xUnholy/ansible-lint-action

Cheers Everyone 👍

**Comment by DO1JLR** [comment]
Hey @xUnholy 

I can verify this PR is [now working](https://github.com/roles-ansible/role_ranger/runs/557299589) for my action lint test.

----

You asked for my opinion about the best readable way to have the cnfig for this action check... 

I like this version:
```yaml
- uses: ansible/ansible-lint-action@master
  with:
    targets: "."
    override-deps: |
      ansible==2.10
      ansible-lint==4.2.0
```
Here it becomes quite obvious to me that versions of pip3 packages are overwritten and which one is affected. Even without looking into the README or the Marketplace description. This is actually not so clear if the setting is really only called "version".
Where I also have some doubts is if it is really such a good idea to write a default version in this repo, because it might be forgotten to update it. And to use the default-pip version unless the individual config explicitly chooses something else is probably not such a problem.

I will still not force an explicit version of ansible-lint for my ansible roles (for now) and therefore it is not as important to me as any optional variables are now called. As long as they remain optional and the functionality of this github action is not compromised.

**Comment by webknjaz** [comment]
@xUnholy you've put a lot of effort into this and I don't want to just disregard it. Thanks for stepping up to do this! But as a maintainer, I want to make sure that it won't cause problems for others and keep it within the project vision.
That said, I'd very much like to accept this PR. So let's agree on the middle ground maybe:
1) All changes to `Dockerfile` should be reverted b/c I still need some version of `ansible-lint` to be there, in the image.
2) Move default to inputs defs

> His opening line says using tags may lead to the same problem? I'm not actually sure what point you're trying to make here?

Ah, I probably should've added more explanation. The point is that people should use commit hashes instead. It's the only safe option here.

**Comment by webknjaz** [comment]
(Oh, I wrote this comment before you pushed two last commits)

File: entrypoint.sh [hunk 13]
```diff
@@ -74,8 +74,12 @@ parse_args() {
   return 0
 }
 
-install_ansible_lint() {
-  pip install ansible-lint=="${VERSION:-${DEFAULT_VERSION}}" || exit 1
+override_python_packages() {
+  echo "Overrides: ${OVERRIDE}"
```
Comment (webknjaz) [inline entrypoint.sh:75]: Other `echo`s in this script write to stderr
```suggestion
  >&2 echo "Overrides: ${OVERRIDE}"
```

File: README.md [hunk 14]
```diff
@@ -35,6 +35,12 @@ jobs:
         #   playbook_2.yml
         targets: ""
         # [optional]
+        # Arguments to override a package and its version to be set explicitly.
+        # Must follow the example syntax.
+        override-dep: |
+          ansible==2.10
```
Comment (webknjaz) [inline README.md:41]: readme shouldn't reference 2.10 b/c it's not released yet. this would confuse people blindly copy-pasting examples:
```suggestion
          ansible==2.9
```

File: README.md [hunk 15]
```diff
@@ -35,6 +35,12 @@ jobs:
         #   playbook_2.yml
         targets: ""
         # [optional]
+        # Arguments to override a package and its version to be set explicitly.
+        # Must follow the example syntax.
+        override-dep: |
```
Comment (webknjaz) [inline README.md:40]: typo
```suggestion
        override-deps: |
```

File: entrypoint.sh [hunk 16]
```diff
@@ -71,6 +71,14 @@ parse_args() {
   return 0
 }
 
+override_python_packages() {
+  echo "Overrides: ${OVERRIDE}"
+  for package in ${OVERRIDE}
```
Comment (webknjaz) [inline entrypoint.sh:76]: It should be possible to do
```suggestion
  pip install ${OVERRIDE}
```
and this is even better because it'll show cross-dep conflicts.

Also, we should probably add
```sh
pip check
```
afterward

File: entrypoint.sh [hunk 17]
```diff
@@ -71,6 +71,10 @@ parse_args() {
   return 0
 }
 
+override_python_packages() {
+  [[ -n ${OVERRIDE} ]] && pip install "${OVERRIDE}"
```
Comment (webknjaz) [inline entrypoint.sh:75]:
```suggestion
  [[ -n ${OVERRIDE} ]] && pip install "${OVERRIDE}" && pip check
```

File: entrypoint.sh (hunk 16 above)
Comment (xUnholy) [reply entrypoint.sh]: How would you like to see the pip check done?

File: entrypoint.sh (hunk 17 above)
Comment (xUnholy) [reply entrypoint.sh:75]: Didn't see this in time for my comment :joy:

**Comment by webknjaz** [comment]
Before proceeding, I need to cherry-pick unrelated changes from this PR into master b/c they're distracting.

File: entrypoint.sh (hunk 8 above)
Comment (webknjaz) [reply entrypoint.sh:57]: I've done some testing and this case matches what it's supposed to match for me. So the removal is wrong.

File: entrypoint.sh [hunk 18]
```diff
@@ -81,13 +85,12 @@ ansible::lint() {
   : "${GITHUB_WORKSPACE?GITHUB_WORKSPACE has to be set. Did you use the actions/checkout action?}"
   pushd "${GITHUB_WORKSPACE}"
 
+  override_python_packages
   local opts
   opts=$(parse_args "$@" || exit 1)
```
Comment (webknjaz) [inline entrypoint.sh:90]:
```suggestion
  opts=$(parse_args "$@" || exit 1)

```

File: entrypoint.sh [hunk 19]
```diff
@@ -81,13 +85,13 @@ ansible::lint() {
...
 
   ansible-lint -v --force-color "$opts" "${TARGETS}"
 }
 
-
 args=("$@")
```
Comment (webknjaz) [inline entrypoint.sh:95]:
```suggestion

args=("$@")
```

File: entrypoint.sh (hunk 8 above)
Comment (webknjaz) [reply entrypoint.sh:57]: Oh, I see what you're doing here

File: action.yml [hunk 20]
```diff
@@ -34,10 +34,16 @@ inputs:
       or valid Ansible directories according to the Ansible role
       directory structure.
     required: true
+  override-deps:
+    description: |
```
Comment (webknjaz) [inline action.yml:38]: Should we also add a default?
```suggestion
    default: ''
    description: |
```

**Comment by DO1JLR** [comment]
currently my github action test is failing: https://github.com/roles-ansible/role_ranger/runs/558326540
The reason for this is #20 and a possible solution is #21 

But with a failing github-action test I will not click at the "approve" button, I guess you will understand that!

**Comment by xUnholy** [comment]
I thought all of my quotations for opts were removed, when did they get into master :eyes:

**Comment by xUnholy** [comment]
@DO1JLR can you run your test again, should be resolved now.

**Comment by webknjaz** [comment]
> I thought all of my quotations for opts were removed, when did they get into master

When I said that I'll cherry-pick things. But I only picked stuff that showed up in the diff in this PR so it wasn't removed.

File: entrypoint.sh [hunk 21]
```diff
@@ -71,6 +71,11 @@ parse_args() {
   return 0
 }
 
+override_python_packages() {
+  [[ -n ${OVERRIDE} ]] && pip install "${OVERRIDE}" && pip check
+  echo "Completed installing override dependencies..."
```
Comment (webknjaz) [inline entrypoint.sh:76]:
```suggestion
  >&2 echo "Completed installing override dependencies..."
```

**Comment by webknjaz** [comment]
Great job, folks!
